target/linux/generic/files/drivers/platform/mikrotik/mikrotik_wlan_lz77.c

```diff
@@ -0,0 +1,444 @@
+					MIKRO_LZ77_MAX_COUNT_BIT_LEN);
+			if (length < 0)
+				return -EINVAL;
+			/* skip bits used by length count */
+			bits_used += bit_count;
+		} else {
+			/* matching group */
+			length = mikrotik_wlan_lz77_decode_count(
+					in,
+					in_offset_bit + bits_used,
+					0, 2, &bit_count,
+					MIKRO_LZ77_MAX_COUNT_BIT_LEN);
+			if (length < 0)
+				return -EINVAL;
+			/* skip bits used by length count */
+			bits_used += bit_count;
+		}
+
+		break;
+
+	case INSTR_ERROR:
+		return -EINVAL;
+	}
+
+	opcode->instruction = instruction;
+	opcode->offset = offset;
+	opcode->length = length;
+	opcode->bits_used = bits_used;
+	opcode->in = (unsigned char *) in;
+	opcode->in_pos = in_offset_bit;
+	return 0;
+}
+
+/**
+ * mikrotik_wlan_lz77_decompress
+ *
+ * @in:			compressed data ptr
+ * @in_len:		length of compressed data
+ * @out:		buffer ptr to decompress into
+ * @out_len:		length of decompressed buffer in input,
+ *			length of decompressed data in success
+ *
+ * Returns 0 on success, or negative error
+ */
+int mikrotik_wlan_lz77_decompress(
+		const unsigned char *in,
+		size_t in_len,
+		unsigned char *out,
+		size_t *out_len)
+{
+	unsigned char *output_ptr;
+	unsigned int input_bit = 0;
+	unsigned char * const output_end = out + *out_len;
+	struct lz77_mk_instr_opcodes *opcode;
+	unsigned int match_offset = 0;
+	int rc = 0;
+	unsigned int match_length, partial_count, i;
+
+	output_ptr = out;
+
+	if (in_len > MIKRO_LZ77_MAX_ENCODED || (in_len * 8) > UINT_MAX) {
+		pr_err(MIKRO_LZ77 "input longer than expected\n");
+		return -EINVAL;
+	}
+
+	opcode = kmalloc(sizeof(struct lz77_mk_instr_opcodes), GFP_KERNEL);
+	if (!opcode)
+		return -ENOMEM;
+
+	while (true) {
+		if (output_ptr > output_end) {
+			pr_err(MIKRO_LZ77 "output overrun\n");
+			goto free_lz77_struct;
+		}
+		if (input_bit > in_len*8) {
+			pr_err(MIKRO_LZ77 "input overrun\n");
+			goto free_lz77_struct;
+		}
+
+		rc = mikrotik_wlan_lz77_decode_instruction_operators(
+				in, input_bit, match_offset, opcode);
+		if (rc < 0) {
+			pr_err(MIKRO_LZ77 "instruction operands decode error\n");
+			goto free_lz77_struct;
+		}
+
+		pr_debug(MIKRO_LZ77 "inbit:0x%x->outbyte:0x%x",
+				input_bit, output_ptr - out);
+
+		input_bit += opcode->bits_used;
+		switch (opcode->instruction) {
+		case INSTR_LITERAL_BYTE:
+			pr_debug(" short");
+			fallthrough;
+		case INSTR_LONG:
+			if (opcode->offset == 0) {
+				/* this is a non-matching group */
+				pr_debug(" non-match, len: 0x%x\n",
+						opcode->length);
+				/* test end marker */
+				if (opcode->length == 0xc &&
+						((input_bit + opcode->length*8) > in_len)) {
+
+					if (*(unsigned int *)out == MIKRO_LZ77_EXPECTED_OUT) {
+						*out_len = output_ptr - out;
+						pr_debug(MIKRO_LZ77 "lz77 decompressed from %zu to %zu\n",
+								in_len, *out_len);
+						return 0;
+					} else {
+						pr_err(MIKRO_LZ77 "lz77 decompressed: unexpected output first word: 0x%08x\n", *(unsigned int *)out);
+						//return -EINVAL;
+						return 0;
```

Comment:
Saw that I was not previously testing first word of decompressed result, so attempted to fix that, and made else case non-fail but noisy; please let me know if you do see this in `dmesg`.

Reply:
Still the same:

```
[    1.112135] [rb_hardconfig] LZ77 tag seen, but decode failed: -2
[    1.112831] [rb_hardconfig] LZ77 tag seen, but decode failed: -2
[    1.117810] [rb_hardconfig] LZ77 tag seen, but decode failed: -2
[    1.123310] MikroTik RouterBOARD hardware configuration sysfs driver v0.08
[    1.132254] MikroTik RouterBOARD software configuration sysfs driver v0.05
```